sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -376,35 +385,76 @@ def get(self, instruction_id, bundle_descriptor_id):
 
     Moves the ``BundleProcessor`` from the inactive to the active cache.
     """
-    try:
-      # pop() is threadsafe
-      processor = self.cached_bundle_processors[bundle_descriptor_id].pop()
-    except IndexError:
-      processor = bundle_processor.BundleProcessor(
-          self.fns[bundle_descriptor_id],
-          self.state_handler_factory.create_state_handler(
-              self.fns[bundle_descriptor_id].state_api_service_descriptor),
-          self.data_channel_factory)
-    self.active_bundle_processors[
+    with self._lock:
+      try:
+        # pop() is threadsafe
+        processor = self.cached_bundle_processors[bundle_descriptor_id].pop()
+        self.active_bundle_processors[
+          instruction_id] = bundle_descriptor_id, processor
+        try:
+          del self.known_not_running_instruction_ids[instruction_id]
```

Comment:
Aren't all instruction ids unique for a given SDK harness/runner combo anyway? If this is the first time we've seen this instruction, why would the instruction be repeated?  I thought only the bundle descriptor ids were re-used?

Reply:
The `instruction_id` refers to the original ProcessBundleRequest `instruction_id` which is different then the current requests `instruction_id` found on `request.instruction_id`

Reply:
Ack. Now that I understand the semantics as the value existing in both before and after processing this makes sense.